frontend/z3_types.py

```diff
@@ -0,0 +1,355 @@
+
+    Dict = type_sort.dict
+    dict_key_type = type_sort.dict_key_type
+    dict_value_type = type_sort.dict_value_type
+
+    Tuple = type_sort.tuple
+    Tuples = get_tuples(type_sort, max_tuple_length)
+
+    Func = type_sort.func
+    Funcs = get_funcs(type_sort, max_function_args)
+
+    Classes = get_classes(type_sort, classes_to_attrs)
+
+    # Encode subtyping relationships
+
+    subtype = Function("subtype", type_sort, type_sort, BoolSort())
+    extends = Function("extends", type_sort, type_sort, BoolSort())
+    not_subtype = Function("not subtype", type_sort, type_sort, BoolSort())
+    stronger_num = Function("stronger num", type_sort, type_sort, BoolSort())
+
+    x = Const("x", type_sort)
+    y = Const("y", type_sort)
+    z = Const("z", type_sort)
+
+    subtype_properties = [
+        ForAll(x, subtype(x, x)),  # reflexivity
+        ForAll([x, y], Implies(extends(x, y), subtype(x, y))),
+        ForAll([x, y, z], Implies(And(subtype(x, y), subtype(y, z)), subtype(x, z))),  # transitivity
+        ForAll([x, y], Implies(And(subtype(x, y), subtype(y, x)), x == y)),
+        ForAll([x, y, z], Implies(And(extends(x, z), extends(y, z), x != y), And(not_subtype(x, y), not_subtype(y, x)))),
+        ForAll([x, y, z], Implies(And(subtype(x, y), not_subtype(y, z)), Not(subtype(x, z)))),
+    ]
+
+    generics_axioms = [
+        ForAll([x, y], Implies(subtype(x, List(y)), x == List(y))),
+        ForAll([x, y], Implies(subtype(x, Set(y)), x == Set(y))),
+        ForAll([x, y, z], Implies(subtype(x, Dict(y, z)), x == Dict(y, z)))
+    ]
+
+    # For numeric casting purposes:
+    # Number > Complex > Float > Int > Bool
+    num_strength_properties = [
+        ForAll(x, Implies(subtype(x, Num), stronger_num(x, x))),  # Reflexivity
+        ForAll([x, y, z], Implies(And(stronger_num(x, y), stronger_num(y, z)), stronger_num(x, z))),  # Transitivity
+        ForAll([x, y], Implies(And(stronger_num(x, y), x != y), Not(stronger_num(y, x)))),
+        ForAll([x, y], Implies(Not(And(subtype(x, Num), subtype(y, Num))),
+                               Not(Or(stronger_num(x, y), stronger_num(y, x)))))
+    ]
+
+    axioms = [
+        extends(zNone, Object),
+        extends(Num, Object),
+        extends(Complex, Num),
+        extends(Float, Num),
+        extends(Int, Num),
+        extends(Bool, Int),
+        extends(Seq, Object),
+        extends(String, Seq),
+        extends(Bytes, Seq),
+        extends(Tuple, Seq),
+
+        extends(Func, Object),
+
+        ForAll([x], extends(Type(x), Object), patterns=[Type(x)]),
+        ForAll([x], extends(List(x), Seq), patterns=[List(x)]),
+        ForAll([x], extends(Set(x), Object), patterns=[Set(x)]),
+        ForAll([x, y], extends(Dict(x, y), Object), patterns=[Dict(x, y)]),
+
+        stronger_num(Int, Bool),
+        stronger_num(Float, Int),
+        stronger_num(Complex, Float),
+        stronger_num(Num, Complex)
+        ] \
+        + tuples_subtype_axioms(Tuples, type_sort) \
```

Comment:
If you wrap the whole block into parentheses, you won't need those backslashes.
axioms = ([ ...
         ], 
         bla,
         blu
)